Fix ordering Transformation for batched dimensions

TimOliverMaier opened the pull request:
This PR is the continuation of @purna135 's PR #5660 addressing issue #5659 (ordering transformation fails for dim>1).

Short summary: 
The initial change of the dimensionality of the jacobian determinant:
```diff
     def log_jac_det(self, value, *inputs): 
-        return at.sum(value[..., 1:], axis=-1)
+        return at.sum(value[..., 1:], axis=-1, keepdims=True)
```
This led to issues with multivariate distributions (as pointed out by @Sayam753 [here](https://github.com/pymc-devs/pymc/pull/5660#discussion_r856940379)).

Based on the [suggestions](https://github.com/pymc-devs/pymc/pull/5660#discussion_r857030059) of @ricardoV94, I added a simple case inside the `log_jac_det` method of `ordered` and `sumto1`transform like this:

```diff
+    def __init__(self, ndim_supp=0):
+        self.ndim_supp = ndim_supp

     def log_jac_det(self, value, *inputs): 
-        return at.sum(value[..., 1:], axis=-1, keepdims=True)
+        if self.ndim_supp == 0:
+            return at.sum(y, axis=-1, keepdims=True)
+        else:
+            return at.sum(y, axis=-1)
```

While fixing the mentioned problem with multivariate distributions, two problems remain:
1.  `TestElementWiseLogP` in `test_transform.py` fails 
     due to the inconsistent dimensionality of the jacobian determinant among the possible transformations. 
     I also ran all other the tests in the `distribution` subfolder locally. Here many tests are also failing in the main branch for me.
     I am unsure here if any of the failing tests are caused by this PR's additions.

2.  While experimenting with the multivariate distributions I realized that ordering of a `MvNormal` fails in main and still fails in 
     this PR. In both branches it fails with an `ValueError`:
     "array must not contain infs or NaNs\nApply node that caused the error: SolveTriangular{lower=True, trans=0, unit_diagonal=False, check_finite=True}(TensorConstant{[[1. 0. 0...0. 0. 1.]]}, SpecifyShape.0)\nToposort index: 12\nInputs types: [TensorType(float64, (4, 4)), TensorType(float64, (4, 1))]\nInputs shapes: [(4, 4), (4, 1)]\nInputs strides: [(8, 32), (8, 8)]\nInputs values: ['not shown', array([[0.80230155],\n       [       nan],\n       [       nan],\n       [       nan]])]\nOutputs clients: [[InplaceDimShuffle{1,0}(SolveTriangular{lower=True, trans=0, unit_diagonal=False, check_finite=True}.0)]]\n\nBacktrace when the node is created (use Aesara flag traceback__limit=N to make it longer):\n  File \"/home/tim/miniconda3/envs/pymc_dev/lib/python3.10/site-packages/aeppl/joint_logprob.py\", line 151, in factorized_joint_logprob\n    q_logprob_vars = _logprob(\n  File \"/home/tim/miniconda3/envs/pymc_dev/lib/python3.10/functools.py\", line 889, in wrapper\n    return dispatch(args[0].__class__)(*args, **kw)\n  File \"/home/tim/miniconda3/envs/pymc_dev/lib/python3.10/site-packages/aeppl/transforms.py\", line 611, in transformed_logprob\n    logprob = _logprob(rv_op, values, *inputs, **kwargs)\n  File \"/home/tim/miniconda3/envs/pymc_dev/lib/python3.10/functools.py\", line 889, in wrapper\n    return dispatch(args[0].__class__)(*args, **kw)\n  File \"/home/tim/Workspaces/pymc/pymc/distributions/distribution.py\", line 123, in logp\n    return class_logp(value, *dist_params)\n  File \"/home/tim/Workspaces/pymc/pymc/distributions/multivariate.py\", line 288, in logp\n    quaddist, logdet, ok = quaddist_parse(value, mu, cov)\n  File \"/home/tim/Workspaces/pymc/pymc/distributions/multivariate.py\", line 156, in quaddist_parse\n    dist, logdet, ok = quaddist_chol(delta, chol_cov)\n  File \"/home/tim/Workspaces/pymc/pymc/distributions/multivariate.py\", line 173, in quaddist_chol\n    delta_trans = solve_lower(chol_cov, delta.T).T\n\nHINT: Use the Aesara flag `exception_verbosity=high` for a debug print-out and storage map footprint of this Apply node."



 For reproduction, see code below.

```python3

with pm.Model() as model:
    kappa_MV = pm.MvNormal("kappa_mv",
                           mu=[1,1,1,1],
                           cov = np.eye(4),
                           initval=[2,1,0,2],
                           transform=pm.distributions.transforms.ordered
                           )
    pm.sample()
```

I want to share my progress here as draft to make collaboration possible ;) . 

Cheers!

**Checklist**
+ [x] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Bugfixes / New features
- ordering transformation for RVs with number of dimensions > 1

purna135 committed:
keep the original dims in the log_jac_det

purna135 committed:
added tests to test_transforms.py

purna135 committed:
added tests for sum_to_1 transforms

purna135 committed:
fix pre-commit

TimOliverMaier committed:
ordered transform class with n_dims_supp

TimOliverMaier committed:
univariate_ordered and multivariate_ordered

TimOliverMaier committed:
ndim_supp cases in sumto1 transform

TimOliverMaier committed:
keep transforms.ordered for backward compatibility

TimOliverMaier committed:
pre-commit run

TimOliverMaier committed:
updated `__all__` variable
1. updated variable `__all__` to contain
   `univariate_ordered`,`multivariate_ordered` and
   analogous `sum_to_1` instantiations.

ricardoV94 commented on pymc/tests/distributions/test_transform.py:
```diff
@@ -598,3 +615,31 @@ def test_discrete_trafo():
         err.match("Transformations for discrete distributions")
+
+
+def test_transforms_ordered():
```
Perhaps a more direct name?

```suggestion
def test_2d_univariate_ordered_():
```

ricardoV94 commented on pymc/tests/distributions/test_transform.py:
```diff
@@ -598,3 +615,31 @@ def test_discrete_trafo():
+            transform=tr.univariate_ordered,
+        )
+
+    log_prob = model.point_logps()
```
To make this test a bit more readable I would suggest including an equivalent `x_1d = pm.Normal(..., shape=(4,))` in the model and comparing the elemwise logp of that is the same as each copy of the 2D `(10, 4)` one that exists now. 

You can do that via `m.compile_logp(sum=False)({"x_1d_ordered__": np.zeros((4,)), x_2d_ordered__": np.zeros(10, 4)})` and then asserting closeness across the last axis.

TimOliverMaier replied:
For the `sum_to_1` transform the logps are not the same. Is this expected? Even if the shape is the same like here:

```python3
def test_2d_univariate_sum_to_1():
    with pm.Model() as model:
        x_1d = pm.Normal(
            "x_1d",
            mu=[-3,-1,1,2],
            sigma=1,
            size=(10,4),
            transform=tr.univariate_sum_to_1,
        )
        x_2d = pm.Normal(
            "x_2d",
            mu=[-3, -1, 1, 2],
            sigma=1,
            size=(10, 4),
            transform=tr.univariate_sum_to_1,
        )

    log_p = model.compile_logp(sum=False)({"x_1d_sumto1__":np.ones((10,3))*0.25,"x_2d_sumto1__":np.zeros((10,3))*0.25})
    np.testing.assert_allclose(log_p[0],log_p[1])
```
Fails with:

```
./pymc/tests/distributions/test_transform.py::test_2d_univariate_sum_to_1 Failed: [undefined]AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 40 / 40 (100%)
Max absolute difference: 1.03125
Max relative difference: 0.72677567
 x: array([[-6.200189, -1.700189, -1.200189, -2.450189],
       [-6.200189, -1.700189, -1.200189, -2.450189],
       [-6.200189, -1.700189, -1.200189, -2.450189],...
 y: array([[-5.418939, -1.418939, -1.418939, -1.418939],
       [-5.418939, -1.418939, -1.418939, -1.418939],
       [-5.418939, -1.418939, -1.418939, -1.418939],...
```

TimOliverMaier replied:
Ok funny. This is not the case if I use `np.zeros`

ricardoV94 replied:
PR looks good otherwise, should we investigate the `SumTo1`?

ricardoV94 replied:
Was this when comparing `np.zeros()`, vs `np.ones()`? I don't expect those to be equivalent...

TimOliverMaier replied:
Oh boy :laughing: . Maybe it was just that. I thought I was comparing `np.ones()` to `np.ones()`. will check again.

TimOliverMaier replied:
Yes. This was my mistake, sorry! Checked it properly with `np.ones()` and it passes.

ricardoV94 reviewed (changes requested):
@TimOliverMaier we forgot to include this PR in our major release of 4.4.0 last night. As such I would agree with your initial suggestion of keeping the original transforms around for a bit longer so we can merge these fixes and release them with 4.1, without breaking backwards compatibility. I hope that's not too annoying :)

Besides that I have some small suggestions for the new tests. I wrote about the first but it applies to both. Let me know what you think

TimOliverMaier committed:
Tests for uni- and multivariate `SumTo1, Ordered`
1. Old `transforms.ordered` and `transforms.sum_to_1` instantiations
   were added again for backwards compatibility.
2. Tests for multivariate and univariate usage of `SumTo1` and `Ordered`
   transformations were added.

TimOliverMaier committed:
test `Ordered` and `SumTo1` raise `ValueError`

TimOliverMaier committed:
improved docs for `SumTo1, Ordered` instantiations
